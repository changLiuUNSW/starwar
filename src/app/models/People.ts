export interface People {
  name: string;
  height: string;
  mass: string;
  birth_year: string;
  gender: string;
  url: string;
  films: string[];
}

export interface PeopleResponse {
  next: string;
  previous: string;
  results: People[];
}
