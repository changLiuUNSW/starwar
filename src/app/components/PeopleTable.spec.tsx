import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { People } from '../models/People';
import PeopleTable from './PeopleTable';

it('PeopleTable renders correctly', () => {
  const items = [
    {
      name: 'Luke Skywalker',
      height: '172',
      mass: '77',
      birth_year: '19BBY',
      gender: 'male',
      url: 'http://swapi.dev/api/people/1/',
      films: [
        'http://swapi.dev/api/films/1/',
        'http://swapi.dev/api/films/2/',
        'http://swapi.dev/api/films/3/',
        'http://swapi.dev/api/films/6/',
      ],
    },
  ] as People[];
  const onRowClick = jest.fn();
  const tree = renderer.create(<PeopleTable items={items} onRowClick={onRowClick} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('PeopleTable should render item correctly', () => {
  const items = [
    {
      name: 'Luke Skywalker',
      height: '172',
      mass: '77',
      birth_year: '19BBY',
      gender: 'male',
      url: 'http://swapi.dev/api/people/1/',
      films: [
        'http://swapi.dev/api/films/1/',
        'http://swapi.dev/api/films/2/',
        'http://swapi.dev/api/films/3/',
        'http://swapi.dev/api/films/6/',
      ],
    },
  ] as People[];
  const onRowClick = jest.fn();
  render(<PeopleTable items={items} onRowClick={onRowClick} />);
  expect(screen.getAllByRole('row')[1].children[0].textContent).toEqual(items[0].name);
  expect(screen.getAllByRole('row')[1].children[1].textContent).toEqual(items[0].height);
  expect(screen.getAllByRole('row')[1].children[2].textContent).toEqual(items[0].mass);
});

it('PeopleTable should trigger row click event when row is clicked', () => {
  const items = [
    {
      name: 'Luke Skywalker',
      height: '172',
      mass: '77',
      birth_year: '19BBY',
      gender: 'male',
      url: 'http://swapi.dev/api/people/1/',
      films: [
        'http://swapi.dev/api/films/1/',
        'http://swapi.dev/api/films/2/',
        'http://swapi.dev/api/films/3/',
        'http://swapi.dev/api/films/6/',
      ],
    },
  ] as People[];
  const onRowClick = jest.fn();
  render(<PeopleTable items={items} onRowClick={onRowClick} />);
  fireEvent.click(screen.getAllByRole('row')[1]);
  expect(onRowClick).toBeCalled();
});

it('PeopleTable should show loading if no items', () => {
  const items = null;
  const onRowClick = jest.fn();
  render(<PeopleTable items={items} onRowClick={onRowClick} />);
  expect(screen.getAllByRole('row')[1].children[0].textContent).toEqual('loading...');
});
