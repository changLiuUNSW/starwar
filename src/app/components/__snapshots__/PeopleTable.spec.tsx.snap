// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeopleTable renders correctly 1`] = `
<div>
  <div
    className="table-responsive"
  >
    <table
      className="table table-hover"
    >
      <thead>
        <tr>
          <th>
            Name
          </th>
          <th>
            Height
          </th>
          <th>
            Mass
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className="sc-bdfBwQ iVZKII"
          onClick={[Function]}
        >
          <td>
            Luke Skywalker
          </td>
          <td>
            172
          </td>
          <td>
            77
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
